### ISSUE TITLE
Question about one time tasks not getting picked

### ISSUE THREAD

**Issue by GowrinathJ** [kagkarlsson/db-scheduler#58]
We are using the scheduler for a huge implementation of one time tasks (typically more than 50 one time tasks every second). After a certain time the db scheduler is not picking up any tasks but seems to work intermittently without consistency. Can you shed some light on the issue we are facing and where we might be going wrong? Also today we observed more than 2500 jobs fail sequential with multiple reattempts and then db scheduler stopped again picking up any tasks

**Comment by kagkarlsson** [comment]
Hi!

Sorry to hear about your troubles! What version are you on? Version before 5.0 had a bug that affected throughput for medium- to high-volume workloads.

**Comment by kagkarlsson** [comment]
The bug was that every `pollingInterval` only a certain number of tasks where executed. It did not run until the queue was empty.

**Comment by kagkarlsson** [comment]
Regarding your last comment, about the failures. I assume the reason for the failure was on your side, but you expected a retry which did not happen? In the latest version, one-time tasks are retried every 5min by default.

**Comment by GowrinathJ** [comment]
We are currently on a 3.X version and I will bump the version to 5.1 to see if that resolves our issue.

**Comment by GowrinathJ** [comment]
Is there an expected thread count that should suffice for the volume we are looking at? we are currently only using 7 threads in total for these scheduler tasks

**Comment by kagkarlsson** [comment]
I am pretty sure 5.1 should solve your issues then :)

Optimal thread-count depends a bit on the type of work the task is doing. 7 should be fine.

**Comment by kagkarlsson** [comment]
If the work is single-threaded and cpu-bound and you have a 16-core box, you might want to go higher :)

**Comment by kagkarlsson** [comment]
or if you are mainly calling out to some high-capacity service, you might also try increasing thread-count

**Comment by GowrinathJ** [comment]
I will try to bump the thread count as well if the version change doesn't seem to do the trick. I got a minor follow up question. Does the scheduler have any wait on the task execution when there are several failures?

**Comment by kagkarlsson** [comment]
Ask away, I enjoy hearing about real-world use-cases :)

There is a default 5min wait per task. That is configurable though (via `FailureHandler`), and it is possible to implement custom retry-behavior, for example one with back-off.

**Comment by kagkarlsson** [comment]
That default might have been different in your version, not sure atm

**Comment by GowrinathJ** [comment]
Sorry, another follow up question as well. On the scheduler_tasks table will the DBScheduler do a constant select on rows? We are facing a deadlock issue now and then when I am trying to retrieve existing tasks. How does the db scheduler handle task execution if the select it does returns a deadlock?

**Comment by kagkarlsson** [comment]
There should be no locking or deadlocks since it uses optimistic locking, i.e. `update ... where task = <task> and version = <last seen version>`. 

What makes you think there is a deadlock?

The scheduler regularly fetches <pollingLimit> due executions from the scheduled_task table. If it fetches the full limit, it should fetch more after it is somewhat done executing the current batch (this only happens in v5.0-).

**Comment by GowrinathJ** [comment]
A simple select is creating a deadlock for me on the tasks table which made me wonder if the same thing happens when the scheduler tries to do a fetch of values. Is there a maximum batch size for the fetch?

**Comment by kagkarlsson** [comment]
What database are you using, and what transaction isolation level are you using for your db-scheduler `Datasource` and for the client you use to run a manual select?

**Comment by GowrinathJ** [comment]
We are using SQL server 2012 but I am not sure about the isolation level for the datasource. Should it be configured to a specific level for the DB Scheduler?

**Comment by GowrinathJ** [comment]
And to update you on the version update, we did bump up the version to 5.1 and still facing intermittent issues. Scheduler is not picking up any tasks for some duration and suddenly starts working and processes all the backlog tasks

**Comment by kagkarlsson** [comment]
I would recommend the default isolation level `READ COMMITTED`. If you set isolation level `SERIALIZABLE` in your client, you might block db-scheduler updates if you select rows in a open transaction. 

It surprises me that 5.1 did not improve your situation, in fact I have a hard time believing it :). Did you see no throughput improvement?

When you say intermittent issues, could you be more specific? How long are the pauses? Are there any warn/errors in the log? Could you catch it with log-level set to `debug` and send me the log?

The behavior of the scheduler is that is wakes every `pollingInterval` (default 30s) and processes all executions that are due. So it will not execute them on the exact `executionTime`.

How many instances of the scheduler are you running btw?

**Comment by onideus** [comment]
Hi @kagkarlsson,

So we've set up a database connection with the read committed over the serializable. We are having intermittent starts without artifact redeploy, it's very peculiar behavior for a single environment. Essentially we'll have a pile of jobs built up, we can still schedule jobs, but the executor is ignoring the jobs that are past execution time. 

Today we had the artifact up and running for a few hours before it started pulling jobs again. There are not any specific warnings or errors in our logs but I can try to set the log-level to debug for db-scheduler.

We're running two instances of the scheduler currently (see below):

@Autowired
	public SchedulerService(@Qualifier("dataSource") DataSource dataSource) {
		oneTimeVodTask = Tasks.oneTime("scheduled-vod-task", ScheduledAlertData.class)
				.execute((inst, ctx) -> {
					videoOnDemandScheduleTask.processVideoOnDemandForEvent(inst.getData().getEventId());
					LOG.info("Executed video on demand task with sensor event id - {}", inst.getData().getEventId());
				});

		oneTimeAlertTask = Tasks.oneTime("scheduled-alert-task", ScheduledAlertData.class)
				.execute((inst, ctx) -> {
					scheduledAlertTaskService.checkEventForAlert(inst.getData().getEventId());
					LOG.info("Executed scheduled alert task with sensor event id - {}", inst.getData().getEventId());
				});

		oneTimeSnapshotTask = Tasks.oneTime("scheduled-snapshot-task", ScheduledAlertData.class)
				.execute((inst, ctx) -> {
					scheduledAlertTaskService.checkEventForSnapshot(inst.getData().getEventId());
					LOG.info("Executed fetching snapshot result task with event id - {}", inst.getData().getEventId());
				});

		oneTimeXmattersTask = Tasks.oneTime("scheduled-xmattersDelete-task", XMattersUserData.class)
				.execute((inst, ctx) -> {
					XMattersService.deleteNumberVerificationTempUserByPersonId(inst.getData().getxMattersPersonId());
					LOG.info("Executed deleting temporary xMatters user with personId " + inst.getData().getxMattersPersonId());
				});

		adhocTaskScheduler = Scheduler
				.create(dataSource, oneTimeVodTask, oneTimeAlertTask, oneTimeSnapshotTask, oneTimeXmattersTask)
				.heartbeatInterval(Duration.ofMinutes(5))
				.threads(5)
				.build();

		Runtime.getRuntime().addShutdownHook(new Thread() {
			@Override
			public void run() {
				LOG.info("Received shutdown signal.");
				adhocTaskScheduler.stop();
			}
		});

		adhocTaskScheduler.start();

		oneTimeNotificationTask = Tasks.oneTime("scheduled-notification-task", ScheduledAlertData.class)
				.execute((inst, ctx) -> {
					scheduledAlertTaskService.checkEventForUserNotification(Long.parseLong(inst.getData().getEventId()), inst.getData().getCount());
					LOG.info("Executed scheduled notification task for user profile id - {}", inst.getData().getEventId());
				});

		notificationTaskScheduler = Scheduler
				.create(dataSource, oneTimeNotificationTask)
				.heartbeatInterval(Duration.ofMinutes(5))
				.threads(2)
				.build();

		notificationTaskScheduler.start();
	}

**Comment by kagkarlsson** [comment]
Thanks for code! Only time for a short answer atm.

If you want to run 2 scheduler-instances in the same jvm, but have a different set of tasks for them, you have to use separate tables for the 2 instances. Override using `builder.tableName("other_tasks")`. Otherwise one instance might mess up for the other if it fetches tasks it has no definitions for.

Also, I would create the tasks as separate beans and have them injected along with the datasource, so I could use the same beans for scheduling new tasks/executions. 

And you should prob. add the 2nd scheduler instance to the shutdown hook.

**Comment by kagkarlsson** [comment]
How are you scheduling new tasks/executions? In the same jvm? Could paste an example of that aswell?

**Comment by onideus** [comment]
Hey @kagkarlsson! Thanks for the rapid response, I didn't get a chance to vpn into our work this weekend, so wanted to touchbase this morning. I have added the second scheduler to the shutdown hook and will be implementing a second table in our hotfix branch (our develop branch actually has 4 schedulers due to the duration that some tasks are taking, so guess I'll be refactoring that as well).

As for the example you requested, here's what is available on the scheduler service class:

`public void scheduleAlertTasks(String eventId, int secondsForFirst, int secondsForSecond, int secondsForThird) {
		LOG.info("Creating first scheduled task for sensor event {}", eventId);

		adhocTaskScheduler.schedule(oneTimeAlertTask.instance(
				eventId + "-first",
				new ScheduledAlertData(eventId)
				),
				Instant.now().plusSeconds(secondsForFirst)
		);

		LOG.info("Creating second scheduled task for sensor event {}", eventId);

		adhocTaskScheduler.schedule(oneTimeAlertTask.instance(
				eventId + "-second",
				new ScheduledAlertData(eventId)
				),
				Instant.now().plusSeconds(secondsForSecond)
		);

		LOG.info("Creating third scheduled task for sensor event {}", eventId);

		adhocTaskScheduler.schedule(oneTimeAlertTask.instance(
				eventId + "-third",
				new ScheduledAlertData(eventId)
				),
				Instant.now().plusSeconds(secondsForThird)
		);
	}

	public void scheduleSnapshotTasks(String eventId) {
		LOG.info("Creating fetching snapshot scheduled task for sensor event {}", eventId);

		adhocTaskScheduler.schedule(oneTimeSnapshotTask.instance(
				"event Id: " + eventId,
				new ScheduledAlertData(eventId)
				),
				Instant.now().plusSeconds(30)
		);
	}

	public void scheduleVodTask(String eventId, int secondsInFuture) {
		LOG.info("Creating VOD scheduled task for sensor event {}", eventId);

		adhocTaskScheduler.schedule(oneTimeVodTask.instance(eventId + "-vod", new ScheduledAlertData(eventId)),
				Instant.now().plusSeconds(secondsInFuture)
		);
	}`

**Comment by onideus** [comment]
Also, additional thought, as the tasks are instantiated with the autowire of the scheduler class constructor with the injected datasource, doesn't that mean that there should only be a single copy? Do I need to separate out the tasks as beans given that the schedulerservice is wiring those up as available and handling the instantiation? I'm not opposed to refactoring those as well if you think it's the right step, but I just am not sure it buys me anything here. Thanks again with all your help sir!

**Comment by kagkarlsson** [comment]
Thanks!

I wanted to make sure that you were using the same instance of the `Task` when you schedule new instances:

```adhocTaskScheduler.schedule(oneTimeAlertTask.instance(```

I assume that `oneTimeAlertTask` is the same task as the one sent to the scheduler on startup. They should be singletons.

**Comment by onideus** [comment]
Correct! They're class level variables handled internally, instead of external classes managing tasks. I figured a scheduler service would handle all tasks as long as the appropriate data was passed along to the scheduler service to be serialized for later use.

	```
private OneTimeTask<ScheduledAlertData> oneTimeVodTask;
	private OneTimeTask<ScheduledAlertData> oneTimeAlertTask;
	private OneTimeTask<ScheduledAlertData> oneTimeSnapshotTask;
	private OneTimeTask<ScheduledAlertData> oneTimeNotificationTask;
```

**Comment by kagkarlsson** [comment]
So the reason you need multiple schedulers is that you want to throttle some tasks, so that they don't occupy all threads and delay other tasks?

**Comment by onideus** [comment]
Correct, we have what we call "vod" tasks that can take as long as a few minutes, but don't have an immediate necessity over our notification and snapshot tasks. So I separated out the schedulers to use different thread pools to ensure that threads were available for our priority tasks.

**Comment by kagkarlsson** [comment]
Also, if you are having problems with one scheduler messing up things for another, you should be gettings these errors in the log:
```LOG.error("Failed to find implementation for task with name '{}'. If there are a high number of executions, this may block other executions and must be fixed.", execution.taskInstance.getTaskName());```

But you are seeing no errors? Logging is working as expected?

**Comment by kagkarlsson** [comment]
Sounds like a good use-case for multiple schedulers :)

**Comment by onideus** [comment]
I adjusted logging and was seeing that log message, correct. I think it's because they're utilizing the same table, as you mentioned previously. I forgot to note that in my earlier message, so your suspicion that our problem was directly related to the schedulers conflicting is likely correct, though it's bizarre that this is exhibited only in our Stage environment (so far).

**Comment by onideus** [comment]
> Sounds like a good use-case for multiple schedulers :)

Glad you like it! Given that you're authoring a framework in use in a variety of ways, I bet it's interesting to hear how people are using (or abusing) your work!

**Comment by kagkarlsson** [comment]
> Glad you like it! Given that you're authoring a framework in use in a variety of ways, I bet it's interesting to hear how people are using (or abusing) your work!

Yeah, definitely! You are however within the limits of its intended use, so I hope we can get to the bottom of what is causing you trouble.

Not entirely sure myself that competing schedulers with different task-set could cause hours of "pause-time" between runs. That sounds strange indeed. And the bug previous to 5.0 would have limited throughput, not blocked it entirely.

After your hotfix, you will also get access to better logging from that environment?

**Comment by kagkarlsson** [comment]
Since you say it is intermittent, it will not be related to any time-zone issues I suppose..

**Comment by kagkarlsson** [comment]
Your threads might be blocked. When you get a "pause" you could try and get a thread-dump and verify that there are available threads in the pool.

**Comment by kagkarlsson** [comment]
Or we can create a diagnostics thread

**Comment by onideus** [comment]
Hey @kagkarlsson, sorry for getting quiet on you. I wanted to make sure that we built all the appropriate tables, specified them, scripted them in flyway, etc. Anyway, I'm seeing successes move through our logs now, and here's our latest schedulerService class (in case you're curious). 

```
@Service
@Configurable
public class SchedulerService {

    public static final Logger LOG = LoggerFactory.getLogger(SchedulerService.class);
    public static final HashMap<Long, Instant> notificationScheduledForUsers = new HashMap<>();

    @Autowired
    @Qualifier("dataSourceScheduler")
    DataSource dataSource;

    @Autowired
    ScheduledAlertTaskService scheduledAlertTaskService;

    @Autowired
    VideoOnDemandScheduleTask videoOnDemandScheduleTask;

    @Autowired
    EventThresholdReport eventThresholdReport;

    @Autowired
    BookkeeperReport bookkeeperReport;

    @Autowired
    XMattersService XMattersService;

    @Autowired
    SettingRepository settingRepository;

    @Autowired
    InfraService infraService;

    @Autowired
    UnacknowledgedReport unacknowledgedReport;


    private final Scheduler notificationTaskScheduler;
    private final Scheduler adhocTaskScheduler;
    private final Scheduler snapshotScheduler;
    private final Scheduler vodScheduler;
    private OneTimeTask<ScheduledAlertData> oneTimeVodTask;
    private OneTimeTask<ScheduledAlertData> oneTimeAlertTask;
    private OneTimeTask<ScheduledAlertData> oneTimeSnapshotTask;
    private OneTimeTask<ScheduledAlertData> oneTimeNotificationTask;
    private OneTimeTask<XMattersUserData> oneTimeXmattersTask;

    @Autowired
    public SchedulerService(@Qualifier("dataSource") DataSource dataSource, SettingRepository settingRepository) {
        oneTimeVodTask = Tasks.oneTime("scheduled-vod-task", ScheduledAlertData.class)
                .execute((inst, ctx) -> {
                    videoOnDemandScheduleTask.processVideoOnDemandForEvent(inst.getData().getEventId());
                    LOG.info("Executed video on demand task with sensor event id - {}", inst.getData().getEventId());
                });

        oneTimeAlertTask = Tasks.oneTime("scheduled-alert-task", ScheduledAlertData.class)
                .execute((inst, ctx) -> {
                    scheduledAlertTaskService.checkEventForAlert(inst.getData().getEventId());
                    LOG.info("Executed scheduled alert task with sensor event id - {}", inst.getData().getEventId());
                });

        oneTimeSnapshotTask = Tasks.oneTime("scheduled-snapshot-task", ScheduledAlertData.class)
                .execute((inst, ctx) -> {
                    scheduledAlertTaskService.checkEventForSnapshot(inst.getData().getEventId());
                    LOG.info("Executed fetching snapshot result task with event id - {}", inst.getData().getEventId());
                });

        oneTimeXmattersTask = Tasks.oneTime("scheduled-xmattersDelete-task", XMattersUserData.class)
                .execute((inst, ctx) -> {
                    XMattersService.deleteNumberVerificationTempUserByPersonId(inst.getData().getxMattersPersonId());
                    LOG.info("Executed deleting temporary xMatters user with personId " + inst.getData().getxMattersPersonId());
                });

        oneTimeNotificationTask = Tasks.oneTime("scheduled-notification-task", ScheduledAlertData.class)
                .execute((inst, ctx) -> {
                    scheduledAlertTaskService.checkEventForUserNotification(Long.parseLong(inst.getData().getEventId()), inst.getData().getCount());
                    LOG.info("Executed scheduled notification task for user profile id - {}", inst.getData().getEventId());
                });

        RecurringTask<Void> dailyBadImagesTask = Tasks.recurring("daily-bad-images-checker-task", FixedDelay.ofHours(24))
                .execute((inst, ctx) -> {
                    LOG.info("Scheduling Bad Images Per Sensor Checker");
                    infraService.infraForBadImage();
                });

        RecurringTask<Void> dailyMissingVodTask = Tasks.recurring("daily-missing-VOD-percent-checker-task", FixedDelay.ofHours(24))
                .execute((inst, ctx) -> {
                    LOG.info("Scheduling Missing VOD Percentage Checker");
                    infraService.getTheRateOfMissingVOD();
                });

        Integer infraNoEventHours = Integer.valueOf(settingRepository.findBySettingNameEquals("INFRA_NO_EVENT_HOURS").getSettingValue());

        RecurringTask<Void> dailySensorEventCheckTask = Tasks.recurring("daily-sensor-event-checker-task", FixedDelay.ofHours(infraNoEventHours))
                .execute((inst, ctx) -> {
                    LOG.info("Scheduling no sensor event checker");
                    infraService.checkNoSensorEvent();
                });

        RecurringTask<Void> weeklyReportTask = Tasks.recurring("weekly-report-task", FixedDelay.ofHours(168))
                .execute((inst, ctx) -> {
                    LOG.info("Scheduling Weekly Reports: Unacknowledged");
                    unacknowledgedReport.createReport();
                });

        RecurringTask<Void> dailyBookkeepeerReportTask = Tasks.recurring("daily-report-task", FixedDelay.ofHours(24))
                .execute((inst, ctx) -> {
                    LOG.info("Scheduling Daily Reports: Event Threshold, Bookkeeper");
                    eventThresholdReport.createReport();
                    bookkeeperReport.createReport();
                });

        adhocTaskScheduler = Scheduler
                .create(dataSource, oneTimeAlertTask, oneTimeXmattersTask, dailyBadImagesTask, dailyMissingVodTask, dailySensorEventCheckTask, weeklyReportTask, dailyBookkeepeerReportTask)
                .heartbeatInterval(Duration.ofMinutes(5))
                .tableName("scheduled_tasks")
                .threads(5)
                .build();

        adhocTaskScheduler.start();

        snapshotScheduler = Scheduler
                .create(dataSource, oneTimeSnapshotTask)
                .heartbeatInterval(Duration.ofMinutes(1))
                .tableName("snapshot_tasks")
                .threads(2)
                .build();

        snapshotScheduler.start();

        vodScheduler = Scheduler
                .create(dataSource, oneTimeVodTask)
                .heartbeatInterval(Duration.ofMinutes(1))
                .tableName("vod_tasks")
                .threads(15)
                .build();

        vodScheduler.start();

        notificationTaskScheduler = Scheduler
                .create(dataSource, oneTimeNotificationTask)
                .heartbeatInterval(Duration.ofMinutes(1))
                .tableName("notification_tasks")
                .threads(2)
                .build();

        notificationTaskScheduler.start();

        Runtime.getRuntime().addShutdownHook(new Thread() {
            @Override
            public void run() {
                LOG.info("Received shutdown signal.");
                adhocTaskScheduler.stop();
                notificationTaskScheduler.stop();
                vodScheduler.stop();
                snapshotScheduler.stop();
            }
        });
    }

    public SchedulerService(DataSource dataSource, ScheduledAlertTaskService scheduledAlertTaskService, SettingRepository settingRepository) {
        this(dataSource, settingRepository);
        this.scheduledAlertTaskService = scheduledAlertTaskService;
    }

    public void scheduleAlertTasks(String eventId, int secondsForFirst, int secondsForSecond, int secondsForThird) {
        if (secondsForFirst == 0) {
            LOG.info("An immediate task was scheduled for sensor event {}", eventId);

            scheduledAlertTaskService.checkEventForAlert(eventId);
        } else {
            LOG.info("Creating first scheduled task for sensor event {}", eventId);

            adhocTaskScheduler.schedule(oneTimeAlertTask.instance(
                    eventId + "-first",
                    new ScheduledAlertData(eventId)
                    ),
                    Instant.now().plusSeconds(secondsForFirst)
            );
        }

        LOG.info("Creating second scheduled task for sensor event {}", eventId);

        adhocTaskScheduler.schedule(oneTimeAlertTask.instance(
                eventId + "-second",
                new ScheduledAlertData(eventId)
                ),
                Instant.now().plusSeconds(secondsForSecond)
        );

        LOG.info("Creating third scheduled task for sensor event {}", eventId);

        adhocTaskScheduler.schedule(oneTimeAlertTask.instance(
                eventId + "-third",
                new ScheduledAlertData(eventId)
                ),
                Instant.now().plusSeconds(secondsForThird)
        );
    }

    public void scheduleSnapshotTasks(String eventId) {
        LOG.info("Creating fetching snapshot scheduled task for sensor event {}", eventId);

        snapshotScheduler.schedule(oneTimeSnapshotTask.instance(
                "event Id: " + eventId,
                new ScheduledAlertData(eventId)
                ),
                Instant.now().plusSeconds(30)
        );
    }

    public void scheduleVodTask(String eventId, int secondsInFuture) {
        LOG.info("Creating VOD scheduled task for sensor event {}", eventId);

        vodScheduler.schedule(oneTimeVodTask.instance(eventId + "-vod", new ScheduledAlertData(eventId)),
                Instant.now().plusSeconds(secondsInFuture)
        );
    }

    public void scheduleUserNotificationTask(Long profileId, int alertCount, int secondsToStart, Instant time) {
        Instant instant = time != null ? time : Instant.now().plusSeconds(secondsToStart);

        LOG.info("UserNotificationTask scheduling at time - {}", Date.from(instant));

        Instant scheduledInstant = notificationScheduledForUsers.get(profileId);

        if (scheduledInstant != null && alertCount > 0) {
            LOG.info("Cancel notification task for user profile id - {}", profileId);
            try {
                notificationTaskScheduler.cancel(
                        TaskInstanceId.of(oneTimeNotificationTask.getName(), String.valueOf(profileId)));
            } catch (Throwable e) {
                LOG.error("An error occurred when attempting to cancel notification for {}", profileId, e);
            }
        } else if (scheduledInstant != null && !scheduledInstant.equals(instant)) {
            LOG.info("Rescheduling notification task for user profile id - {}", profileId);
            try {
                notificationTaskScheduler.reschedule(
                        TaskInstanceId.of(oneTimeNotificationTask.getName(), String.valueOf(profileId)), instant);
                LOG.info("Rescheduled notification task at time - {}", Date.from(instant));
                return;
            } catch (Throwable e) {
                LOG.error("An error occurred when attempting to rescheduled notification for {}", profileId, e);
            }
        }

        notificationScheduledForUsers.put(profileId, instant);

        LOG.info("Creating scheduled notification task for user profile id - {}", profileId);

        notificationTaskScheduler.schedule(oneTimeNotificationTask.instance(
                String.valueOf(profileId),
                new ScheduledAlertData(String.valueOf(profileId), alertCount)
        ), instant);

        LOG.info("Created scheduled notification task at time - {}", Date.from(instant));
    }

    public void scheduleXMattersTempUserDeletion(String xMattersPersonId, Instant time) {
        LOG.info("Creating xMatters Temp User deletion task for personId :" + xMattersPersonId);

        adhocTaskScheduler.schedule(oneTimeXmattersTask.instance("xMattersPersonId: " + xMattersPersonId, new XMattersUserData(xMattersPersonId)), time);
    }
}
```

Anyway, we're going to be monitoring the environment to make sure that this is the fix that we needed it to be, so fingers crossed! Thanks again for all your help, and I will keep you posted either way (so you can see what we're doing with your framework if nothing else!).

**Comment by kagkarlsson** [comment]
Pleased to see the use of many of the different features that have been added over time :). Let me know if you feel you are missing some important feature, especially if it would be easy to add.

And keep me posted on the result of the fix 👍

**Comment by kagkarlsson** [comment]
How did this work out for you?

**Comment by onideus** [comment]
Hey @kagkarlsson!

So it looks like I have good and bad news. The good news is that the fix you suggested worked very well. The bad news is it seems to have worked too well, and the throughput that we gained from this newly managed scheduler is causing strain on other portions of our codebase. In addition, we discovered an issue with our application where when we allow the Spring Boot framework to determine thread pool size for the executor that it's passing to our event listener that is attached to an IBM MQ, when it sees your framework getting loads of threads (which is what we want and need for the purposes of our use case) it pulls the reins hard on the threadpool for the listener.

As such, I had to dig into the way that we were crafting our listener so that it would have enough threads to handle the load of messages that we're required to consume for our application. Once that was corrected, we started having JPA locks due to the size as well as the new conflict with the number of simultaneous threads jockying for lock position from not just our listener but also the scheduler. So we had to craft a means of processing batches of messages to reduce lock times and loads on our JPA intermediary layer for data access.

Anyway, long story short, your help proved invaluable in getting our scheduler up and running in an extremely efficient manner in a single application, and has been much appreciated. I apologize I didn't reply sooner with the results, as it had a cascade effect that was unanticipated and had to be dealt with.

Thanks again for all that you've done!

**Comment by kagkarlsson** [comment]
Thanks for the thorough report! Out of curiosity, what throughput did you achieve?

**Comment by onideus** [comment]
With regards to your scheduler, we're working roughly 300 scheduled tasks a minute right now (and look to have plenty of room to grow), some of which require more than a few seconds of work. We're only in approximately 30 satellite locations right now, with the intention of running through the events from more than 2700 satellite locations once to scale. Ultimately I think we'll break the scheduler off into its own microservice to allow for parallel growth of additional instances when the load swells, using our cloud platform to manage simultaneous containers. However, for today these changes provide exactly what we needed to move forward with our work.

**Comment by kagkarlsson** [comment]
Closing this discussion :)

**Comment by kagkarlsson** [comment]
🎉 This issue has been resolved in `v16.9.0` ([Release Notes](https://github.com/kagkarlsson/db-scheduler/releases/tag/v16.9.0))
